cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UsageHandler.java

```diff
@@ -0,0 +1,260 @@
+    final Id.Stream id = Id.Stream.from(namespaceId, streamId);
+    Set<? extends Id> ids = executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageDatasetIterable, Set<? extends Id>>() {
+        @Override
+        public Set<? extends Id> apply(UsageDatasetIterable input) throws Exception {
+          return input.getUsageDataset().getPrograms(id);
+        }
+      });
+    responder.sendJson(HttpResponseStatus.OK, ids);
+  }
+
+  @GET
+  @Path("/namespaces/{namespace-id}/streams/{stream-id}/adapters")
+  public void getStreamAdapterUsage(HttpRequest request, HttpResponder responder,
+                                    @PathParam("namespace-id") String namespaceId,
+                                    @PathParam("stream-id") String streamId) {
+    final Id.Stream id = Id.Stream.from(namespaceId, streamId);
+    Set<? extends Id> ids = executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageDatasetIterable, Set<? extends Id>>() {
+      @Override
+      public Set<? extends Id> apply(UsageDatasetIterable input) throws Exception {
+        return input.getUsageDataset().getAdapters(id);
+      }
+    });
+    responder.sendJson(HttpResponseStatus.OK, ids);
+  }
+
+  @GET
+  @Path("/namespaces/{namespace-id}/data/datasets/{dataset-id}/programs")
+  public void getDatasetAppUsage(HttpRequest request, HttpResponder responder,
+                                 @PathParam("namespace-id") String namespaceId,
+                                 @PathParam("dataset-id") String datasetId) {
+    final Id.DatasetInstance id = Id.DatasetInstance.from(namespaceId, datasetId);
+    Set<? extends Id> ids = executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageDatasetIterable, Set<? extends Id>>() {
+        @Override
+        public Set<? extends Id> apply(UsageDatasetIterable input) throws Exception {
+          return input.getUsageDataset().getPrograms(id);
+        }
+      });
+    responder.sendJson(HttpResponseStatus.OK, ids);
+  }
+
+  @GET
+  @Path("/namespaces/{namespace-id}/data/datasets/{dataset-id}/adapters")
+  public void getDatasetAdapterUsage(HttpRequest request, HttpResponder responder,
+                                     @PathParam("namespace-id") String namespaceId,
+                                     @PathParam("dataset-id") String datasetId) {
+    final Id.DatasetInstance id = Id.DatasetInstance.from(namespaceId, datasetId);
+    Set<? extends Id> ids = executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageDatasetIterable, Set<? extends Id>>() {
+        @Override
+        public Set<? extends Id> apply(UsageDatasetIterable input) throws Exception {
+          return input.getUsageDataset().getAdapters(id);
+        }
+      });
+    responder.sendJson(HttpResponseStatus.OK, ids);
+  }
+
+  @VisibleForTesting
+  public <R> R executeUsageDatasetOp(TransactionExecutor.Function<UsageDatasetIterable, R> func) {
```

Comment:
Needs to be public right now b/c for some reason all the handler tests are in a different package..